wgengine/netstack/link_endpoint.go

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package netstack
+
+import (
+	"context"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/tcpip"
+	"gvisor.dev/gvisor/pkg/tcpip/header"
+	"gvisor.dev/gvisor/pkg/tcpip/stack"
+)
+
+type queue struct {
+	c      chan *stack.PacketBuffer
```

Comment:
If we're wrapping access to `c` in a `RWMutex`, I wonder if it would be cheaper to have a linked list or ring buffer protected by a mutex, which I think is basically what a channel is anyway. Maybe it's still needed for the blocking reads.

Reply:
Added a TODO in https://github.com/tailscale/tailscale/pull/12869/commits/016fbfed2d63c614642d4d7b8c652b6ff26c65f6. I'm tempted to glue the write path directly into the packet handling TUN path and avoid an intermediary buffer entirely, but it requires more thought on the impacts of that back pressure.